freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/backup/cloud/CloudBackupConfigGenerator.java

```diff
@@ -0,0 +1,37 @@
+package com.sequenceiq.freeipa.service.freeipa.backup.cloud;
+
+import java.nio.file.Paths;
+
+import org.apache.commons.lang3.StringUtils;
+
+public abstract class CloudBackupConfigGenerator<T extends CloudBackupStorageConfig> {
+
+    protected static final String CLUSTER_BACKUP_PREFIX = "cluster-backups";
+
+    public abstract String generateBackupLocation(String location, String clusterType,
+            String clusterName, String clusterId);
+
+    String getLocationWithoutSchemePrefixes(String input, String... schemePrefixes) {
```

Comment:
im wondering why is this needed here. I had to split and store different parts of the storage location parts (scheme + host + path) because of fluentd configuration requirements, i think in your case the path will work as is as you are using s3 tool for the upload (probably same is true for azure)...so probably most of these path generator code are not required at all...just pass the storage location as is (with appending the cluster-backup and cluster name/id prefixes)

Reply:
Actually this needs the exact same sort of segmentation as logs.  Adding additional user-defined folders as well as segmenting the into the per cluster based folders is all still necessary.

Reply:
An example of this is right now in out SDX environments we have a shared bucket which we put in the environment name as the top-level under the bucket in order to allow us to put the logs and data both under that bucket on a per-environment basis

Reply:
but is the splitting really needed? concat segments would be enough instead of the generator service usage as schema/host parts can be included in (at least) s3 tool (so transformations can be avoided)